gcd/maestro_gcd.go

```diff
@@ -0,0 +1,278 @@
+/*
+Copyright (c) 2020, Arm Limited and affiliates.
+SPDX-License-Identifier: Apache-2.0
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package gcd
+
+
+import (
+	"encoding/json"
+	"os"
+	"errors"
+	"io/ioutil"
+	"github.com/armPelionEdge/maestro/log"
+	"github.com/armPelionEdge/maestro/maestroConfig"
+	b64 "encoding/base64"
+)
+
+type gwResourceManagerRegisterArgs struct {
+	Name string `json:"name"`
+}
+
+type addResouceArgs struct {
+	Lwm2mObjects []lwm2mObject `json:"objects"`
+}
+
+type lwm2mObject struct {
+	ObjectId int 					 `json:"objectId"`
+	ObjectInstances []objectInstance `json:"objectInstances"`
+}
+
+type objectInstance struct {
+	ObjectInstanceId int `json:"objectInstanceId"`
+	Resources []resource `json:"resources"`
+}
+
+type resource struct {
+	ResourceId int  `json:"resourceId"`
+	Operations int  `json:"operations"`
+	Type string `json:"type"`
+	Value string    `json:"value"`
+}
+
+var grm_client *Client = nil   // Gateway Resource Manager Client
+var gcd_config *maestroConfig.GcdConfig // Gateway Capability Discovery config
+
+
+
+func Gcd_init(config *maestroConfig.GcdConfig) {
+
+	if(config == nil) {
```

Comment:
> (required) What happens when gateway_capabilities is not included in maestro_config.yaml? Need to protect backwards compatibility with previous versions of the maestro config

@mray190, looks like that concern is addressed here.  The Gcd_init function checks for nil and returns immediately.  @ygoyal18 I'm not sure this should be logged as an error though.

Reply:
@costanic Agreed, will change that.